### PR TITLE
Fix.python

### PR DESCRIPTION
**Pull request by dawa79**
performance  improvement

### CODE REVIEW

File: scripts/fregion.py [hunk 1]
```diff
@@ -9,11 +9,19 @@
 #
 #    to read the stored field 'x' out of f:
 #      f.x
-#
+       
+#    If x is an array type, it is lazy loaded, and it supports the interface as following:
+#      1) for d in f.x()
```
Comment (kthielen) [inline scripts/fregion.py]: For this case would it be possible to make it 'for d in f.x'?  It's probably just a question of supporting that 'getitems' function or whatever it is right?

File: scripts/fregion.py [hunk 2]
```diff
@@ -28,13 +36,69 @@
 #      and:
 #        the returned value may be whatever the application decides is sensible
 #
+#    2. Just use the decorator RegReader like:
```
Comment (kthielen) [inline scripts/fregion.py]: Nice!

File: scripts/fregion.py [hunk 3]
```diff
@@ -542,7 +606,7 @@ def V1toV2Type(ty):
   tyDisp = {
     "prim":    lambda p:  p if (p.rep == None) else Prim(p.name, V1toV2Type(p.rep)),
     "var":     lambda v:  v,
-    "farr":    lambda fa: FixedArr(V1toV2Type(fa.ty), V1toV2Type(ta.tlen)),
+    "farr":    lambda fa: FixedArr(V1toV2Type(fa.ty), V1toV2Type(fa.tlen)),
```
Comment (kthielen) [inline scripts/fregion.py]: Oops!  Good catch!  >_<

File: scripts/fregion.py [hunk 4]
```diff
@@ -745,7 +809,8 @@ def __init__(self, renv, fs, tys):
   def read(self,m,offset):
     vs=[]
     for i in range(len(self.os)):
-      vs.append(self.rs[i].read(m,offset+self.os[i]))
+      vs.append(Loader(self.rs[i].read, m, offset+self.os[i]))
+      #vs.append(self.rs[i].read(m,offset+self.os[i]))
```
Comment (kthielen) [inline scripts/fregion.py]: Maybe if we don't need this line, just delete it?

File: scripts/fregion.py [hunk 5]
```diff
@@ -804,7 +869,9 @@ def __init__(self, renv, ctors):
     self.cns  = cns
   def read(self,m,offset):
     t = self.tr.read(m,offset)
-    return VariantView(self.cns[t], self.crs[t].read(m,offset+self.poff))
+    v = Loader(self.crs[t].read, m, offset+self.poff)
+    return VariantView(self.cns[t], v)
+    #return VariantView(self.cns[t], self.crs[t].read(m,offset+self.poff))
```
Comment (kthielen) [inline scripts/fregion.py]: Same here, don't need this line right?

File: scripts/fregion.py (hunk 1 above)
Comment (dawa79) [reply scripts/fregion.py]: yes, you can . you can see the class RecStream, it is just a wrapper of f.x

File: scripts/fregion.py [hunk 6]
```diff
@@ -813,19 +880,59 @@ def read(self,m,offset):
...
 class ArrReader:
   def __init__(self,renv,ty):
     self.nr   = UnpackReader('Q',8)
     self.r    = makeReader(renv,ty)
     self.vlen = sizeOf(ty)
+
   def read(self,m,offset):
+    n=self.nr.read(m,offset)
+    return ArrReaderGenerator(m, self, n, offset+8)
+    """
```
Comment (kthielen) [inline scripts/fregion.py]: I think that this block-comments out this code?

File: scripts/fregion.py (hunk 1 above)
Comment (kthielen) [reply scripts/fregion.py]: In that case maybe the comment should be updated to say that you can write it as 'for d in f.x()'?  Just a little simpler from the API perspective.

File: scripts/fregion.py [hunk 7]
```diff
@@ -1015,41 +1122,44 @@ def meta(f): return FRMeta(f)
...
 #######
 
+def RegReader(desc):
+  def newCls(cls):
+    FRegion.addType(desc, lambda renv, ty, repty: cls(renv, ty, repty))
+    return cls
+  return newCls
+
 # date/time
+@RegReader("datetime")
```
Comment (kthielen) [inline scripts/fregion.py]: This is great, simpler to read as a self-contained definition.

Comment (dawa79) [reply scripts/fregion.py]: Yes. that is python. I really hope c++ can add this feature in future

File: test/Python.C [hunk 8]
```diff
@@ -227,14 +229,14 @@ def smapInto(r,f,x):
...
-    r.append(f(x[0]))
+    r.append(f(x[0]()))
     return smapInto(r,f,x[1])
 def smap(f,x):
   r=[]
   smapInto(r,f,x)
   return r
 
-if (sum(smap(lambda vs: sum(map(lambda v:v.x, vs)),f.stss)) != 207900):
+if (sum(smap(lambda vs: sum(map(lambda v:v.x, vs)),f.stss[1])) != 207900):
```
Comment (kthielen) [inline test/Python.C]: Could you explain why this change had to be made?

Comment (dawa79) [reply test/Python.C]: just fixed. now it is f.stss

File: scripts/fregion.py [hunk 9]
```diff
@@ -1177,29 +1172,29 @@ def __getitem__(self,k):
     if (self.sl.count==0):
       return None
     else:
-      n = SLView.findNextGLEB(self.sl.root, len(self.sl.root.next)-1, k)
-      if (not(n == None) and n.key==k):
-        return n.value
+      n = SLView.findNextGLEB(self.sl.root(), len(self.sl.root().next.value)-1, k)
```
[the remark this answers is not in the document]
Comment (dawa79) [reply scripts/fregion.py]: Yes I agree. I will working on that

[the remark this answers is not in the document]
Comment (dawa79) [reply scripts/fregion.py]: for type struct and variant we can keep it like x.f
for the tuple,removed the lazy load.
